grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java

```diff
@@ -981,28 +1051,108 @@ public GrpcService build() {
             grpcExceptionHandler = exceptionHandler;
         }
 
+        // We will copy the service and methods inside the old registry to new one and intercept them.
+        final HandlerRegistry.Builder newRegistryBuilder = new HandlerRegistry.Builder();
+
         if (grpcExceptionHandler != null) {
-            registryBuilder.setDefaultExceptionHandler(grpcExceptionHandler);
+            newRegistryBuilder.setDefaultExceptionHandler(grpcExceptionHandler);
         }
 
-        if (interceptors != null) {
```

Comment:
How about configuring the interceptors within the `HandlerRegistry` and constructing the service within it? This would enable us to encapsulate the logic more effectively.
```java
public class GrpcServiceBuilder {
    public GrpcServiceBuilder intercept(Iterable<? extends ServerInterceptor> interceptors) {
        requireNonNull(interceptors, "interceptors");
        registryBuilder.addInterceptors(interceptors);
        return this;
    }
    ..
}
final class HandlerRegistry {
    ...
    HandlerRegistry build() {
        // interceptors are used for building services.
    }
}
```

Reply:
I have initially tried this. I was unsuccessful, let me retry.

Reply:
The complex part is that we need to rebuild the handler registry. I previously tried to modify `Builder.build()` method. I think I will handle this instead as a separate method rather than trying to intercept everything last minute inside `build()`.

Reply:
I moved this logic to the builder and I modified entries right before we iterate over them.